pallets/multisig-verifier/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+        NotTrustedPeerSignature,
+        NoSuchPeer,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig {
+        /// Network id for current network
+        pub network_id: GenericNetworkId,
+    }
+
+    #[cfg(feature = "std")]
+    impl Default for GenesisConfig {
+        fn default() -> Self {
+            Self {
+                network_id: GenericNetworkId::Sub(SubNetworkId::Mainnet),
+            }
+        }
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config> GenesisBuild<T> for GenesisConfig {
+        fn build(&self) {
+            ThisNetworkId::<T>::put(self.network_id);
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(0)]
+        pub fn initialize(
+            origin: OriginFor<T>,
+            network_id: GenericNetworkId,
+            keys: Vec<ecdsa::Public>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+            ensure!(keys.len() > 0, Error::<T>::InvalidInitParams);
+
+            let btree_keys = keys.into_iter().collect();
+            PeerKeys::<T>::set(network_id, Some(btree_keys));
+            Self::deposit_event(Event::NetworkInitialized(network_id));
+            Ok(().into())
+        }
+
+        #[pallet::call_index(1)]
+        #[pallet::weight(0)]
+        pub fn add_peer(
+            origin: OriginFor<T>,
+            network_id: GenericNetworkId,
+            key: ecdsa::Public,
+        ) -> DispatchResultWithPostInfo {
+            let output = T::CallOrigin::ensure_origin(origin)?;
+            frame_support::log::info!(
+                "Call add_peer {:?} by {:?}",
+                key,
+                output
+            );
+            PeerKeys::<T>::try_mutate(network_id,|x| -> DispatchResult {
+                let Some(keys) = x else {
+                    fail!(Error::<T>::NetworkNotInitialized)
+                };
+                keys.insert(key);
+                Ok(())
+            })?;
+            Self::deposit_event(Event::PeerAdded(key));
+            Ok(().into())
+        }
+
+        #[pallet::call_index(2)]
+        #[pallet::weight(0)]
+        pub fn remove_peer(
+            origin: OriginFor<T>,
+            network_id: GenericNetworkId,
+            key: ecdsa::Public,
+        ) -> DispatchResultWithPostInfo {
+            let output = T::CallOrigin::ensure_origin(origin)?;
+            frame_support::log::info!(
+                "Call remove_peer {:?} by {:?}",
+                key,
+                output
+            );
+            PeerKeys::<T>::try_mutate(network_id,|x| -> DispatchResult {
+                let Some(keys) = x else {
+                    fail!(Error::<T>::NetworkNotInitialized)
+                };
+                ensure!(keys.remove(&key), {
+                    Self::deposit_event(Event::NoSuchPeer(key));
```

Comment:
I'm not sure if it will work because by default all calls wrapped is transactional and this event will be removed from storage if you return error

Reply:
It makes sense, I will check